pkg/flow/declare_test.go

```diff
@@ -0,0 +1,308 @@
+			}
+		
+			test "myModule" {
+				input = testcomponents.count.inc.count
+			}
+		
+			testcomponents.summation "sum" {
+				input = test.myModule.output
+			}
+			`,
+			expected: 10,
+		},
+		{
+			name: "DeclaredInParentDepth1",
+			config: `
+			declare "test" {
+				argument "input" {
+					optional = false
+				}
+			
+				testcomponents.passthrough "pt" {
+					input = argument.input.value
+					lag = "1ms"
+				}
+				rootDeclare "default" {
+					input = testcomponents.passthrough.pt.output
+				}
+			
+				export "output" {
+					value = rootDeclare.default.output
+				}
+			}
+			declare "rootDeclare" {
+				argument "input" {
+					optional = false
+				}
+				export "output" {
+					value = argument.input.value
+				}
+			}
+			testcomponents.count "inc" {
+				frequency = "10ms"
+				max = 10
+			}
+		
+			test "myModule" {
+				input = testcomponents.count.inc.count
+			}
+		
+			testcomponents.summation "sum" {
+				input = test.myModule.output
+			}
+			`,
+			expected: 10,
+		},
+		{
+			name: "DeclaredInParentDepth2",
+			config: `
+			declare "test" {
+				argument "input" {
+					optional = false
+				}
+			
+				testcomponents.passthrough "pt" {
+					input = argument.input.value
+					lag = "1ms"
+				}
+				declare "anotherDeclare" {
+					argument "input" {
+						optional = false
+					}
+					rootDeclare "default" {
+						input = argument.input.value
+					}
+					export "output" {
+						value = rootDeclare.default.output
+					}
+				}
+				anotherDeclare "myOtherDeclare" {
+					input = testcomponents.passthrough.pt.output
+				}
+			
+				export "output" {
+					value = anotherDeclare.myOtherDeclare.output
+				}
+			}
+			declare "rootDeclare" {
+				argument "input" {
+					optional = false
+				}
+				export "output" {
+					value = argument.input.value
+				}
+			}
+			testcomponents.count "inc" {
+				frequency = "10ms"
+				max = 10
+			}
+		
+			test "myModule" {
+				input = testcomponents.count.inc.count
+			}
+		
+			testcomponents.summation "sum" {
+				input = test.myModule.output
+			}
+			`,
+			expected: 10,
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			ctrl := flow.New(testOptions(t))
+			f, err := flow.ParseSource(t.Name(), []byte(tc.config))
+			require.NoError(t, err)
+			require.NotNil(t, f)
+
+			err = ctrl.LoadSource(f, nil)
+			require.NoError(t, err)
+
+			ctx, cancel := context.WithCancel(context.Background())
+			done := make(chan struct{})
+			go func() {
+				ctrl.Run(ctx)
+				close(done)
+			}()
+			defer func() {
+				cancel()
+				<-done
+			}()
+
+			require.Eventually(t, func() bool {
+				export := getExport[testcomponents.SummationExports](t, ctrl, "", "testcomponents.summation.sum")
+				return export.LastAdded == tc.expected
+			}, 3*time.Second, 10*time.Millisecond)
+		})
+	}
+}
+
+type errorTestCase struct {
+	name          string
+	config        string
+	expectedError *regexp.Regexp
+}
+
+func TestDeclareError(t *testing.T) {
```

Comment:
I'd like to see test cases for at least:

* Ensuring that a custom component can't reference top-level declares, since its being scoped out for now. 

* Ensuring that a nested declare can't access components out of scope:

  ```river
  declare "a" {
    b_1 "default" { } // this should error 
  }
  declare "b" {
    declare "b_1" {}
  }
  ```

* What happens when you declare a component called `declare`? (IMO this should be invalid, and should probably be a reserved name alongside import.)

If you can come up with any other test cases for where declare can/should fail, please add them as well.

Reply:
> Ensuring that a custom component can't reference top-level declares, since its being scoped out for now.

What do you mean by this?
The following case is allowed:
```
declare "a" {}
declare "b" { a {} }
b {}
```
Or do you mean that it should not be supported to reference the exports of a node declare outside of the declare?

Reply:
Thanks for the suggestions, although the code logic matches the expected outcome, the error messages are not ideal for the tests: 
- OutOfScopeReference: error: component "testcomponents.count.inc.count" does not exist
```
       testcomponents.count "inc" {
		frequency = "10ms"
		max = 10
	}

	declare "example_a" {
		testcomponents.summation "sum" {
			input = testcomponents.count.inc.count // invalid reference
		}
	}
	example_a "test" {}
```
- OutOfScopeDefinition: error: unrecognized component name "b_1"
```
	declare "a" {
		b_1 "default" { } // this should error 
	}
	declare "b" {
		declare "b_1" {}
	}
	a "example" {}
```
In both cases, the declare cannot tell the difference between a reference that does not exist or a reference that is out of scope because it does not have access to the full scope. 
I suggest that we either keep it like this, or we change the messages to smth like "component "testcomponents.count.inc.count" does not exist or is out of scope".

Reply:
@wildum the new error message you proposed ("X does not exist or is out of scope") makes sense to me, let's move to that 👍

Reply:
@wildum my gut feeling for that one is to change up the error message a little bit to something like:

> component name X is not defined in this scope

I'm not sure why this is my gut feeling; I think I want to have the phrasing different to more easily differentiate between errors for invalid references and invalid declarations

Reply:
users that only use built-in components and are not aware of custom components may find this message a bit obscure if they just mistyped the name of a component. In this context it's not very clear what "scope" refers to.
Maybe "cannot retrieve the definition of component name X"?

Reply:
`cannot retrieve the definition of component name X` works for me, :shipit: